### DISCUSSION TITLE
Changing data frames to tibbles

### DISCUSSION THREAD

**Discussion by lilyclements** [IDEMSInternational/R-Instat#9765] [General]
Following a recent discussion, I have been looking into changing data frames to tibbles, and the impact this may have.

First I'll outline general issues seen, and then the problems found when I implemented it in R-Instat and ran through the "Prepare" menu.

**General Issues**
I have an R-file [here](https://github.com/africanmathsinitiative/R-Instat/files/5108652/Tibbles.vs.Data.Frames.zip) that runs through the general issues outlined in this section.

1. tibbles do not use row names, however, data frames do.
There is an option in the function `as_tibble` to add the rowname as a row with the parameter `rownames` (taking a string for the name of the new row).
E.g.
```
mtcars.tib <- as_tibble(mtcars, rownames = "Car")
mtcars.df <- data.frame(mtcars)
```
If we convert data frames to tibbles, can we have this option as a checkbox in the import page? (And I suppose also in the "From Library" page too for data sets such as `mtcars`).

2. Default class for strings change between tibbles and data frames.
In tibbles, they are read in as characters. In data frames, as factors.
```
x = c("A", "B", "C", "D")

tib.test <- as_tibble(x)
class(tib.test$value) # character

df.test <- as.data.frame(x)
class(df.test$x) # factor
```

3. Using square brackets usually is the same between them, except for `[ , . ]` 
```
mtcars.tib[,1] # does the same as []   i.e., first column as tibble
mtcars.df[,1] # does the same as [[]]  i.e., first column as vector
```
As stated [here](https://www.jumpingrivers.com/blog/the-trouble-with-tibbles/#tibbles-vs-data-frames) this can bring in an issue with older packages: 
"Some older packages don't work with tibbles because of their alternative subsetting method. They expect tib[,1] to return a vector, when infact it will now return another tibble."

**Running in R-Instat**
Running through the "Prepare" menu with a tibble rather than data frame has brought up four points.

1. The use of the dialog `Data Frame > Row Numbers/names`. 
Tibbles do not have row names and so the three options here are not relevant?
a) Copy Row Names to Column - not needed if the row names are just 1-nrow(data)
b) Set Row Names from Column - this function does not do anything with tibbles
c) Reset Row Names to Positive Integers - not needed if the row names are just 1-nrow(data)

2. `Error in FUN(left): invalid argument type`
This error seems to occur if a new column is added to the data set and it is not added to the end of the data set, but instead in the middle. Dialogs I have found this occur with:
_a) Data Frame > Row Numbers/names - when selecting the radio button to copy row names to the (first) column
(although if we have tibbles, this function is not needed?)
b) Data Frame > Filter - when making a new data frame
c) Data Frame > Reorder Columns
d) Data Frame > Insert Columns/Rows - error unless a new column is inserted at the end of the data set.
e) Data Frame > Delete Columns/Rows - error only if deleting rows.
f) Column: Calculate > Duplicate - error (provided the final column is not selected to be duplicated)
g) Column: Text > Find/Replace - error (provided the final column is not selected to be duplicated)
h) Data Reshape > Append Data Frames_

3. `Error in if (!tibble::as_tibble(new_data)) { : missing value where TRUE/FALSE needed`
This error seems to occur when creating a new data set from the current tibble. Dialogs (in the Prepare menu) where this occurs are:
_a) Column: Factor > Factor Data Frame
b) Data Reshape > Stack (with mtcars, I just put one variable in)
c) Data Reshape > Unstack
d) Data Reshape > Transpose
e) Data Reshape > Random Subset_
 
4. `Error in matrix(if (is.null(value)) logical() else value, nrow = nr, dimnames = list(rn,  : 
  length of 'dimnames' [2] not equal to array extent`
This is the only place in the Prepare menu where I found this occurs.
Take the mtcars data, convert `cyl` to a character, and open this dialog. Put in `cyl` as the character column, and check the checkbox `Create Subset ...`
This error does not occur with a data frame.
I think it is to do with having a 1x0 tibble.
If I add "hi" as any value in the `cyl` vector then the dialog works fine (although, does bring up the error seen in 1.).

I can continue going through other menus to find any further issues.

**Comment by shadrackkibet** [comment]
Just to point out that if we are going to move soon into R version `4..` series,  then the default behaviour of  `as.dataframe()` function and other subsequent functions eg `read.csv()` that used to coarse strings into factors have changed. see the changes [here](https://cran.r-project.org/doc/manuals/r-devel/NEWS.html). So your point (2) above will both return type `character` by default in R verison `4..`

(3) If we want tibbles to return a vector when one column is selected like ordinary dataframes then setting `drop=TRUE` will suffice. So the code below will both return a vector.

```
mtcars.tib <- as_tibble(mtcars)
mtcars.df <- data.frame(mtcars)

mtcars.tib[,1, drop = TRUE] 
mtcars.df[,1]
```
